backend/src/models/db.js

```diff
@@ -1,125 +1,229 @@
-    (err) => {
-        if (err) {
-            return console.log(err.message)
-        }
-        console.log('Successfully added test data')
+}
+
+function createReplyTable (db) {
+    db.run(`CREATE TABLE reply (
+        answerID INTEGER PRIMARY KEY AUTOINCREMENT,
+        reviewID INTEGER NOT NULL,
+        userID INTEGER NOT NULL,
+        body TEXT NOT NULL,
+        timestamp TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+        likes INTEGER DEFAULT '0.00',
+        FOREIGN KEY (reviewID) REFERENCES reviews(courseID),
+        FOREIGN KEY (userID) REFERENCES users(userID))`
+    )
+}
+
+exports.createDB = function (databaseName) {
+    // Create the database object
+    const db = new sqlite3.Database(databaseName)
+
+    // Create the database tables
+    db.serialize(() => {
+        createUserTable(db)
+        createUniversityTable(db)
+        createCourseTable(db)
+        createQuestionTable(db)
+        createAnswerTable(db)
+        createReviewTable(db)
+        createReplyTable(db)
+    })
+
+    // If in memory databse, intialise it
+    if (databaseName === ':memory:') {
+        exports.devInitDB(db)
+    }
+
+    return db
+}
+
+// Insert given JSON object into database table.
+// data = { column : value }
+// For security reasons, column inputs can NEVER be user defined.
+// TODO - FIX ISSUE WHERE this.lastID != user/uni/course/question/answerID
+//        THIS IS DUE TO THE FACT THAT WE CAN ONLY RETURN THE ROW_ID AFTER
+//        INSERTION WITHOUT A LOOKUP... TOO TIRED TO SOLVE RN
+exports.insertDB = function (db, table, data) {
+    return new Promise((resolve, reject) => {
+        const columns = Object.keys(data)
+        const placeholders = columns.map((column) => '?').join()
+        const query = `INSERT INTO ${table} (${columns}) VALUES (${placeholders})`
+        console.log(query)
+        db.run(
+            query,
+            Object.values(data),
+            function (err, row) { err ? reject(err) : resolve(this.lastID) }
+        )
+    })
+}
+
+exports.devInitDB = function (db) {
+    /* Fake data objects */
+    const user = {
+        firstName: 'Walker',
+        lastName: 'Francis',
+        email: 'alnuno-das-hinds@gmail.com',
+        password: 'ilovetravxoxo'
+    }
+
+    const unsw = {
+        universityName: 'Univerity of New South Wales'
     }
+
+    const comp4920 = {
+        courseCode: 'COMP4920',
+        courseName: 'Ethics and Management',
+        facultyCode: 'COMP'
+    }
+
+    const question = {
+        title: 'Question creation help!',
+        body: 'I am struggling to create a meme question, wud shud I rite?'
+    }
+
+    const answer = {
+        body: 'Knock. Knock... Who is there?... Pivotal... Pivotal Who?... lol jks it is Trello!'
+    }
+
+    const review = {
+        body: 'Would bang out of 10!'
+    }
+
+    const reply = {
+        body: 'Good to know... *** sick bastard ***'
+    }
+
+    /* Insert the fake data in to the database */
+    const promises = []
```

Comment:
This seems quite odd. Is this because some updates depend on the next? If the updates are required to be synchronous chain the promises. `Promise.all()` is if the updates both need to be done for some data transform. Also simplifies error handling; you can just follow with a single catch.

Reply:
Yeah so things like a 'question' depends on there being a 'user' and a 'course', then an 'answer' depends on there being a 'user' and a 'question'. Just did some hacky system to make it synchronous and access all the past promises. Wasn't sure how to do it cleanly because you can only get the previous promises', result when chaining (not the ones before that)?